src/tests/end2end/DepthStencilStateTests.cpp

```diff
@@ -0,0 +1,512 @@
+            RGBA8 greaterColor = RGBA8(0, 0, 255, 255);
+
+            TestSpec base = { baseState, baseColor, 0.0f, 1u };
+            DoTest({ base, { state, lessColor, 0.f, 0u } }, less ? lessColor : baseColor);
+            DoTest({ base, { state, equalColor, 0.f, 1u } }, equal ? equalColor : baseColor);
+            DoTest({ base, { state, greaterColor, 0.f, 2u } }, greater ? greaterColor : baseColor);
+        }
+
+        // Wipe the stencil value to `initialStencil` then draw a triangle such that is always passes the depth and stencil tests with the provided reference
+        // Check that the stencil value is as expected
+        void CheckStencilOperation(nxt::StencilOperation stencilOperation, uint32_t initialStencil, uint32_t reference, uint32_t expectedStencil) {
+            nxt::DepthStencilState baseState = device.CreateDepthStencilStateBuilder()
+                .SetStencilFunction(nxt::Face::Both, nxt::CompareFunction::Always, nxt::StencilOperation::Keep, nxt::StencilOperation::Keep, nxt::StencilOperation::Replace)
+                .GetResult();
+
+            nxt::DepthStencilState state = device.CreateDepthStencilStateBuilder()
+                .SetStencilFunction(nxt::Face::Both, nxt::CompareFunction::Always, nxt::StencilOperation::Keep, nxt::StencilOperation::Keep, stencilOperation)
+                .GetResult();
+
+            CheckStencil({
+                { baseState, RGBA8(255, 255, 255, 255), 0.f, initialStencil },
+                { state, RGBA8(255, 0, 0, 255), 0.f,  reference },
+            }, expectedStencil);
+        }
+
+        // Given a list of test specs, add an additional test that always draws a green triangle if the stencil value is equal to the expected value
+        void CheckStencil(const std::vector<TestSpec> &testParams, uint32_t expectedStencil) {
+            nxt::DepthStencilState state = device.CreateDepthStencilStateBuilder()
+                .SetStencilFunction(nxt::Face::Both, nxt::CompareFunction::Equal, nxt::StencilOperation::Keep, nxt::StencilOperation::Keep, nxt::StencilOperation::Keep)
+                .GetResult();
+
+            std::vector<TestSpec> params(testParams.begin(), testParams.end());
+            params.push_back({ state, RGBA8(0, 255, 0, 255), 0, expectedStencil });
+            DoTest(params, RGBA8(0, 255, 0, 255));
+        }
+
+        // Each test param represents a triangle with a different pipeline state. Create a separate pipeline and bind group for each triangle.
+        // Populate the bind group's buffer view with the depth and color data for each triangle.
+        // Create the render pipeline for each triangle
+        // For each triangle, set the triangle data, set the stencil reference, and draw
+        void DoTest(const std::vector<TestSpec> &testParams, const RGBA8& expectedFront, const RGBA8& expectedBack) {
+            nxt::CommandBufferBuilder builder = device.CreateCommandBufferBuilder();
+
+            struct TriangleData {
```

Comment:
The TriangleData structure could be put on the stack directly inside the loop that iterates on the specs.